### PR TITLE
Component refactoring

### PR DESCRIPTION
**Pull request by emulbreh**

### CODE REVIEW

File: lymph/core/components.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        if factory in self._declared_components:
+            raise RuntimeError("already installed: %s" % factory)
+        component = factory(self, **kwargs)
+        self._declared_components[factory] = component
+        self.add_component(component)
+        return component
+
+    def on_start(self):
+        for declaration in self.declarations:
+            declaration.__get__(self, type(self))
```
Comment (mamachanko) [inline lymph/core/components.py]: Isn't there a more "natural" way to do this?

File: lymph/core/components.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            if isinstance(value, Declaration):
+                value.name = name
+                declarations.add(value)
+        new_cls = super(ComponentizedBase, cls).__new__(cls, clsname, bases, attrs)
+        new_cls.declarations = declarations
+        return new_cls
+
+
+@six.add_metaclass(ComponentizedBase)
+class Componentized(Component):
```
Comment (mamachanko) [inline lymph/core/components.py]: Not a fan of `Componentized`, but will not complain I can think of something better. On the other hand it goes well together with `Component`.

File: lymph/core/components.py (hunk 1 above)
Comment (emulbreh) [inline lymph/core/components.py]: I should have added a WIP comment, this code will go away.

File: lymph/core/container.py [hunk 3]
```diff
@@ -56,14 +57,18 @@ def __init__(self, ip='127.0.0.1', port=None, registry=None, logger=None, events
...
-        self.install(DefaultInterface, interface_name='lymph')
+        self.add_component(registry)
         registry.install(self)
         if events:
+            self.add_component(events)
             events.install(self)
 
+        self.server = self.install(self.server_cls, ip=ip, port=port)
+        self.monitor = self.install(MonitorPusher, endpoint=monitor_endpoint, interval=5)
+        self.install_interface(DefaultInterface, name='lymph')
```
Comment (mamachanko) [inline lymph/core/container.py]: The way components get added to and installed on each other is a little confusing. Throwing in `install_interface` doesn't make things easier to grok.

File: docs/api/components_api.rst [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    
+        Adds `component`.
+    
+    .. attribute:: metrics
+    
+        A :class:`Metrics` instance that aggregates the metrics from all added components.
+    
+    .. method:: on_start()
+    
+        Calls `on_start()` on all added components.
```
Comment (mamachanko) [inline docs/api/components_api.rst]: Shouldn't it be instead?

_"Should call `on_start()` on all added components."_

File: lymph/core/monitoring/pusher.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        self.loop_greenlet.kill()
+
+    def loop(self):
+        last_stats = time.monotonic()
+        while True:
+            gevent.sleep(self.interval)
+            dt = time.monotonic() - last_stats
+            series = list(self.aggregate)
+            stats = {
+                'dt': dt,
```
Comment (mamachanko) [inline lymph/core/monitoring/pusher.py]: delta?

File: docs/api/components_api.rst (hunk 4 above)
Comment (emulbreh) [inline docs/api/components_api.rst]: It *does* call `on_start()`, i.e. it's not an interface but a base class.

File: lymph/core/rpc.py [hunk 6]
```diff
@@ -217,13 +220,7 @@ def _recv_loop(self):
...
-                'requests': dict(self.request_counts)
-            },
-            'connections': [c.stats() for c in self.connections.values()],
-        }
-        self.request_counts.clear()
-        return stats
+    def raw_metrics(self):
+        yield 'rpc.connection_count', len(self.connections), {}
+        for subject, count in self.request_counts.items():
+            yield 'rpc.request_count', count, {'subject': subject}
```
Comment (mamachanko) [inline lymph/core/rpc.py]: removed `self.request_counts.clear()` on purpose?

File: docs/api/components_api.rst (hunk 4 above)
Comment (mamachanko) [inline docs/api/components_api.rst]: understood

File: lymph/core/monitoring/pusher.py (hunk 5 above)
Comment (emulbreh) [inline lymph/core/monitoring/pusher.py]: Then it'd have to be `delta_t` (or time_delta). This is just keeping the data we had in the stats before, but I think we can remove it completely now.

File: lymph/core/rpc.py (hunk 6 above)
Comment (emulbreh) [inline lymph/core/rpc.py]: yes, we should monitor absolutes instead of rates.

File: lymph/core/components.py (hunk 2 above)
Comment (manisenkov) [inline lymph/core/components.py]: `Compound` or `Composite`? Just came into my mind

**Comment by mamachanko** [comment]
heavy stuff! nice.

lgtm

**Pull request merged**